src/dataframe/dataframe.jl

```diff
@@ -1266,15 +1266,17 @@ function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; cols::S
         end
         return df
     end
-    if cols === :orderequal
+    if cols == :orderequal
         if row isa NamedTuple
-            if any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
+            if length(row) != ncol(df) || any(x -> x[1] != x[2], zip(propertynames(row), _names(df)))
                 throw(ArgumentError("when `cols=:orderequal` all data frames must have " *
                                     "the same column names and in the same order"))
             end
         end
-    elseif cols === :setequal || cols === :equal
-        if cols === :equal
+    end
+    # for AbstractDict :orderequal is allowed ant treated as :setequal
```

Comment:
This will preclude dicts that have an order (like OrderedDict and SortedDict). There also is no easy opt-in/opt-out. I think it is a fair decision to allow failing on unordered collections if `cols=:orderequal` is supplied. I'd not default to ignoring order if it is explicitely specified (remember that `:setequal` is the default. Thus, `:orderequal` will only appear if explicitely required)